Google.Api.Gax.Grpc/Rest/RestMethod.cs

```diff
@@ -82,19 +78,124 @@ internal static RestMethod Create(MethodDescriptor method, JsonParser parser)
         /// <returns>A request with the URI, method and content populated.</returns>
         internal HttpRequestMessage CreateRequest(IMessage protoRequest, string host)
         {
-            string path = _pathPattern.Format(protoRequest);
-            var uri = host is null ? new Uri(path, UriKind.Relative) : new UriBuilder { Host = host, Path = path }.Uri;
-            string jsonContent = _contentFactory?.Invoke(protoRequest);
+            var transcodingResult = Transcode(protoRequest);
+
+            var uriPathWithParams = AppendQueryStringParameters(transcodingResult.UriPath, transcodingResult.QueryStringParameters.OrderBy(kvp => kvp.Key));
+
+            var uri = host is null ? new Uri(uriPathWithParams, UriKind.Relative) : new UriBuilder { Host = host, Path = transcodingResult.UriPath }.Uri;
+            
             return new HttpRequestMessage
             {
                 RequestUri = uri,
                 Method = _httpMethod,
-                Content = jsonContent is null ? null : new StringContent(jsonContent, Encoding.UTF8, s_applicationJsonMediaType)
+                Content = transcodingResult.Body is null ? null : new StringContent(transcodingResult.Body, Encoding.UTF8, s_applicationJsonMediaType),
             };
         }
 
-        // TODO: Handle cancellation?
+        /// <summary>
+        /// Merges the uri path and the query string parameters, escaping them.
+        /// Ignores the possibility that the path can already have parameters or contain an anchor (`#`)
+        /// </summary>
+        /// <param name="uriPath">The path component of the service URI</param>
+        /// <param name="queryStringParameters">The parameters to encode in the query string</param>
+        /// <returns></returns>
+        private static string AppendQueryStringParameters(string uriPath, IOrderedEnumerable<KeyValuePair<string, string>> queryStringParameters)
+        {
+            var sb = new StringBuilder();
+            sb.Append(uriPath);
+            queryStringParameters.Select((kvpNameValue, i) => new
```

Comment:
I suspect this would be simpler (and more efficient) as a foreach loop. I'm happy for it to go in like this and I can create a PR with a suggestion (that can be closed if you prefer this) rather than try to describe exactly what I mean or write a suggestion without validating it.

Reply:
👍